compiler/plc_index/src/lib.rs

```diff
@@ -0,0 +1,72 @@
+use encoding_rs::Encoding;
+use plc_ast::provider::IdProvider;
+use plc_source::source_location::SourceLocation;
+use plc_source::{SourceCode, SourceContainer};
+use std::collections::HashMap;
+
+#[derive(Debug, Default)]
+pub struct GlobalContext {
+    sources: HashMap<&'static str, SourceCode>,
+    provider: IdProvider,
+    // TODO: The following would be also nice, to have a cleaner API i.e. instead of working with different structs such
+    //       as the index or the diagnostics one could instead ONLY use the `GlobalContext` with methods like
+    //       `ctxt.{add,get}_diagnostics(...)` making the code perhaps a bit cleaner / reducing the # of arguments for
+    //       some functions / methods?
+    //       RefCells may or may not make sense here, because maybe we dont want to pass the GlobalContext as a mutable reference?
+    //       -> diagnostics: RefCell<Diagnostics>, (private visibility)
+    //       -> index: RefCell<Index>, (private visibility; `get_index(&self) -> &mut Index`?)
+}
+
+impl GlobalContext {
+    pub fn new() -> Self {
+        Self { sources: HashMap::new(), provider: IdProvider::default() }
+    }
+
+    pub fn with_source<S: SourceContainer>(mut self, sources: &[S], enc: Option<&'static Encoding>) -> Self {
+        sources.iter().for_each(|source| self.insert(source, enc));
+        self
+    }
+
+    pub fn insert(&mut self, container: &impl SourceContainer, encoding: Option<&'static Encoding>) {
+        if let Ok(value) = container.load_source(encoding) {
+            self.sources.insert(container.get_location_str(), value);
+        }
```

Comment:
This will silently not populate the HashMap upon error, does returning a `Result` make sense here?

Reply:
Yes, i would prefer results